src/core/providers/gdal/qgsgdalprovider.cpp

```diff
@@ -4622,6 +4622,10 @@ int QgsGdalProviderMetadata::listStyles( const QString &uri, QStringList &ids, Q
   }
   QVariantMap uriParts = QgsGdalProviderBase::decodeGdalUri( uri );
   QString layerName = uriParts["layerName"].toString();
+  if ( layerName.isEmpty() )
+  {
+    layerName = GDALGetMetadataItem( ds.get(), "IDENTIFIER", "" );
+  }
```

Comment:
doesn't that beg to be factored into a small helper function, like ``getLayerNameForStyle(const QString& uri, gdal::dataset_unique_ptr& ds)`` ?

Reply:
> is it also possible to add it to QgsGdalProviderBase::decodeGdalUri?

I would probably restrict using IDENTIFIER as the layer name to GeoPackage only. There could be other formats where there would be a custom IDENTIFIER metadata item with perhaps some UUID or non end-user friendly content.

Reply:
Good to know. In that case, I'm going for the helper function.